Implement ticklabelposition for cartesian subplots and colorbars

archmoj opened the pull request:
Resolves #5186 by implement `ticklabelposition` for cartesian subplots as well as `colorbars`.
and fixes #5262 (in c7a23b4a1da099136ecada926d5297c631884984)
and fixes #5301 (in 6e80dc98a14e4c895a586763ae12aa8280f2cb25, 8efa7bd668af97ac1e9c6ccd200219e72df7823a and 6f575a3e8f85b9cc855f86328b48fca082fbbb05) 

`ticklabelposition` determines where tick labels are drawn with respect to the axis. 
Options are
`outside`, `inside`,
`outside top`, `inside top`,
`outside left`, `inside left`,
`outside right`, `inside right`,
`outside bottom`, `inside bottom`.
      
It Has no effect on *multicategory* axes or when `tickson` is set *boundaries*.

Please note that
top or bottom has no effect on x axes or when `ticklabelmode` is set to *period*.
Similarly,
left or right has no effect on y axes or when `ticklabelmode` is set to *period*.

[Demo](https://codepen.io/MojtabaSamimi/pen/dyXLGOg?editors=1000)

@plotly/plotly_js

nicolaskruchten commented:
To check:
* category axes
* period positioning

nicolaskruchten commented:
recapping conversation in slack: it's pretty much a core part of the intention of this feature that we take into account the bounding boxes of the labels, rather than hardcoding multiples of the font-height.

archmoj commented:
> recapping conversation in slack: it's pretty much a core part of the intention of this feature that we take into account the bounding boxes of the labels, rather than hardcoding multiples of the font-height.

Addressed in https://github.com/plotly/plotly.js/pull/5275/commits/c262d4bff6df5caab93fcce286b53113c422ede1 and https://github.com/plotly/plotly.js/pull/5275/commits/74f594def71f65a3f012c3f066fe469bc9f3471e.

alexcjohnson commented:
Looking great! Nice job with the extra autorange, and with removing labels that extend outside the plot. A few things I notice playing with the demo:
- If I zoom/pan the y axis and change the label length while the x axis is still autoranged, I would have expected the x axis range to recalculate to fit the new labels but it doesn't. Somehow if I double-click TWICE on the x axis at that point it will recalculate, but I'd have wanted it to happen automatically. Here for example is starting with short labels (eg 8k) and making them longer by zooming (8000):
<img width="308" alt="Screen Shot 2020-11-22 at 8 36 56 AM" src="https://user-images.githubusercontent.com/2678795/99905356-53f5ba80-2c9e-11eb-921d-f4e36029e1a9.png">

- Sometimes as I'm interacting with the plot, the x axis _data_ will appear to lose the extra range for the y labels, but the x axis _ticks and gridlines_ will be correct. I can't detect a clear pattern for when this happens, and sometimes I see just a small shift, not all the way to where the data would have been without the inside labels. This looks similar to a bug I encountered in #5287 and fixed with an extra `ax.setScale` at the [right place in the process](https://github.com/plotly/plotly.js/pull/5287/files#diff-091de2c0f84c090af5b656d3aeeca4acb530b82e5f99afb367be575da138455eR26)
<img width="368" alt="Screen Shot 2020-11-22 at 8 34 11 AM" src="https://user-images.githubusercontent.com/2678795/99905504-29583180-2c9f-11eb-8291-bb1f19bf9fa3.png">

alexcjohnson commented on src/plots/cartesian/axis_defaults.js:
```diff
@@ -117,6 +125,38 @@ module.exports = function handleAxisDefaults(containerIn, containerOut, coerce,
+            containerOut.ticklabelposition =
+            containerOut.ticklabelposition
+                .replace(' top', '')
+                .replace(' bottom', '');
```
The pattern we've normally used for this kind of situation is to change the available options based on other factors. See for example what we did with `side`:

https://github.com/plotly/plotly.js/blob/432e448161797aeeffe06a26e4a8ab2ae98524ec/src/plots/cartesian/position_defaults.js#L50-L56

This means we wouldn't try to keep part of the user's intention as you do here (currently "inside" would survive, even if "top" didn't) but I think that's better, plain "inside" will probably look bad when you meant "inside right" but you actually said "inside top".

archmoj replied:
Thanks for the note.
Addressed in https://github.com/plotly/plotly.js/pull/5275/commits/2361958804bebee2c789382367976ad2517f0ca2.
Resolving...

alexcjohnson commented on src/plots/cartesian/layout_attributes.js:
```diff
@@ -505,6 +505,27 @@ module.exports = {
+            'top or bottom has no effect on x axes or when `ticklabelmode` is set to *period*.',
+            'Similarly',
+            'left or right has no effect on y axes or when `ticklabelmode` is set to *period*.',
+            'Has no effect on *multicategory* axes or when `tickson` is set *boundaries*.'
```
```suggestion
            'Has no effect on *multicategory* axes or when `tickson` is set to *boundaries*.'
```

alexcjohnson commented on src/plots/plots.js:
```diff
@@ -2060,8 +2060,20 @@ plots.doAutoMargin = function(gd) {
+    for(var i = 0; i < axList.length; i++) {
+        var ax = axList[i];
+
+        var hideFn = ax._hideOutOfRangeInsideTickLabels;
```
While dragging to pan or zoom I still see the out-of-range labels, but then on mouse up they disappear. Not a big deal, but it would be a bit nicer to never see them.

archmoj replied:
Addressed in https://github.com/plotly/plotly.js/pull/5275/commits/47c57738c3a52f0ba4c0790cf7a7032e040d1c62.
Resolving...

alexcjohnson commented on test/image/mocks/ticklabelposition-1.json:
```diff
@@ -0,0 +1,149 @@
+            "ticklabelmode": "period",
+            "side": "bottom",
+            "ticks": "inside",
+            "ticklabelposition": "inside",
```
These labels look a little off-center?
<img width="278" alt="Screen Shot 2020-11-22 at 9 17 48 AM" src="https://user-images.githubusercontent.com/2678795/99906182-a8e7ff80-2ca3-11eb-82aa-355bd219f86c.png">

archmoj replied:
Please note that the mode of labels on this axis is *period*. So the labels are not centered on the ticks.
Resolving...

alexcjohnson commented on test/jasmine/tests/mock_test.js:
```diff
@@ -959,6 +959,9 @@ var list = [
     'tickformatstops',
+    'ticklabelposition-0',
+    'ticklabelposition-1',
+    'ticklabelposition-2',
```
Oh I forgot about this step in #5287 - would you mind adding `axes_chain_scaleanchor_matches` and `axes_chain_scaleanchor_matches2` for me?

archmoj replied:
Done in https://github.com/plotly/plotly.js/pull/5275/commits/91ea2f616a092189d78c2334b5602b0327507dab.

nicolaskruchten commented:
Here's a case we could do a bit better in: https://codepen.io/nicolaskruchten/pen/WNGebgW?editors=0010
* the xaxis tick label overlaps the first point
* the DDDDD yaxis tick label is hidden

archmoj commented:
> Here's a case we could do a bit better in: https://codepen.io/nicolaskruchten/pen/WNGebgW?editors=0010
> 
> * the xaxis tick label overlaps the first point
> * the DDDDD yaxis tick label is hidden

Now renders slightly better as illustrated in [this codepen](https://codepen.io/MojtabaSamimi/pen/xxEKmpZ?editors=0010) and as a result of https://github.com/plotly/plotly.js/pull/5275/commits/70e708a4a2c67b3a8a3413cb9b0358b3223fba4b.

archmoj commented:
> Looking great! Nice job with the extra autorange, and with removing labels that extend outside the plot. A few things I notice playing with the demo:
> 
> * If I zoom/pan the y axis and change the label length while the x axis is still autoranged, I would have expected the x axis range to recalculate to fit the new labels but it doesn't. Somehow if I double-click TWICE on the x axis at that point it will recalculate, but I'd have wanted it to happen automatically. Here for example is starting with short labels (eg 8k) and making them longer by zooming (8000):
> 
> <img alt="Screen Shot 2020-11-22 at 8 36 56 AM" width="308" src="https://user-images.githubusercontent.com/2678795/99905356-53f5ba80-2c9e-11eb-921d-f4e36029e1a9.png">
> 
> * Sometimes as I'm interacting with the plot, the x axis _data_ will appear to lose the extra range for the y labels, but the x axis _ticks and gridlines_ will be correct. I can't detect a clear pattern for when this happens, and sometimes I see just a small shift, not all the way to where the data would have been without the inside labels. This looks similar to a bug I encountered in #5287 and fixed with an extra `ax.setScale` at the [right place in the process](https://github.com/plotly/plotly.js/pull/5287/files#diff-091de2c0f84c090af5b656d3aeeca4acb530b82e5f99afb367be575da138455eR26)
> 
> <img alt="Screen Shot 2020-11-22 at 8 34 11 AM" width="368" src="https://user-images.githubusercontent.com/2678795/99905504-29583180-2c9f-11eb-8291-bb1f19bf9fa3.png">

Good catch. Fixed by https://github.com/plotly/plotly.js/pull/5275/commits/999f87776ba5a0ede3a980c821c7430594b50c52.

archmoj commented:
> Looking great! Nice job with the extra autorange, and with removing labels that extend outside the plot. A few things I notice playing with the demo:
> 
> * If I zoom/pan the y axis and change the label length while the x axis is still autoranged, I would have expected the x axis range to recalculate to fit the new labels but it doesn't. Somehow if I double-click TWICE on the x axis at that point it will recalculate, but I'd have wanted it to happen automatically. Here for example is starting with short labels (eg 8k) and making them longer by zooming (8000):
> 
> <img alt="Screen Shot 2020-11-22 at 8 36 56 AM" width="308" src="https://user-images.githubusercontent.com/2678795/99905356-53f5ba80-2c9e-11eb-921d-f4e36029e1a9.png">

Good call. Addressed in https://github.com/plotly/plotly.js/pull/5275/commits/445eee4ee21cadef37981e57db63619ef6d12b66.

alexcjohnson commented on src/plots/cartesian/axes.js:
```diff
@@ -2467,36 +2472,80 @@ axes.getTickSigns = function(ax) {
     return d.periodX !== undefined ? d.periodX : d.x;
 }
 
+function getTickLabelUV(ax) {
```
minor point - u & v don't mean much to me. The function makes sense, just might be worth a comment like "u is a shift along the axis, v is a shift perpendicular to the axis" for future reference.

archmoj replied:
Good call. Added in https://github.com/plotly/plotly.js/pull/5275/commits/40c09b78daba4ddabb2ff4bbed9b2536783bae29.

alexcjohnson commented on src/plots/cartesian/axes.js:
```diff
@@ -2887,6 +3029,40 @@ axes.drawLabels = function(gd, ax, opts) {
+                        else if(bb.top + (ax.tickangle ? 0 : d.fontSize / 4) < min) hide = true;
+                    }
+                    if(hide) thisLabel.select('text').style({ opacity: 0 });
+                } // TODO: hide mathjax?
```
heh inside labels and MathJax hopefully have rather disjoint user sets... So yes we should test this at some point, but not necessary for this PR

alexcjohnson reviewed (approved):
Fantastic! Nice job including colorbars and improving various other aspects of rotated tick label positioning. 💃
